### PR TITLE
Xray package, mostly a copy of v2ray, but with better UDP support

### PR DESCRIPTION
**Pull request by iopq**
###### Description of changes

Added xray as a package and a service

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/networking/xray/default.nix [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+    for file in $out/bin/*; do
+      makeWrapper "$file" "$out/bin/$(basename "$file")" \
+        --set-default XRAY_LOCATION_ASSET ${assetsDrv}/share/xray
+    done
+  '';
+
+  meta = {
+    homepage = "https://github.com/XTLS/Xray-core";
+    description = "Xray, Penetrates Everything. Also the best v2ray-core, with XTLS support and fully compatible configuration";
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/xray/default.nix:50]:
```suggestion
    description = "Xray, Penetrates Everything. Also the best v2ray-core, with XTLS support and fully compatible configuration";
    homepage = "https://github.com/XTLS/Xray-core";
```

File: pkgs/tools/networking/xray/default.nix [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  nativeBuildInputs = [ makeWrapper ];
+
+  doCheck = false;
+
+  buildPhase = ''
+    buildFlagsArray=(-v -p $NIX_BUILD_CORES -ldflags="-s -w -buildid=")
+    runHook preBuild
+    go build "''${buildFlagsArray[@]}" -o xray ./main
+    runHook postBuild
+  '';
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/xray/default.nix:37]: Can we use the standard buildPhase?

File: pkgs/tools/networking/xray/default.nix [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  assetsDrv = symlinkJoin {
+    name = "v2ray-assets";
+    paths = assets;
+  };
+
+in
+buildGoModule rec {
+  pname = "v2ray";
+  inherit version src;
+  inherit vendorSha256;
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/xray/default.nix:26]:
```suggestion
let
  assetsDrv = symlinkJoin {
    name = "v2ray-assets";
    paths = assets;
  };

in
buildGoModule rec {
  pname = "v2ray";
  version = "1.5.5";

  src = fetchFromGitHub {
    owner = "XTLS";
    repo = "Xray-core";
    rev = "v${version}";
    sha256 = "x2aNAu+H3qJIKjQbE0rYxfQAWucvOMaU2eSy9YIZdcQ=";
  };

  vendorSha256 = "sha256-PBcUKUP06QKTu3djxoTb7B7Ld86fZz7+SDewRZy1YmQ=";
```

File: nixos/modules/services/networking/xray.nix [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        type = types.package;
+        default = pkgs.xray;
+        defaultText = literalExpression "pkgs.xray";
+        description = lib.mdDoc ''
+          Which xray package to use.
+        '';
+      };
+
+      configFile = mkOption {
+        type = types.nullOr types.str;
```
Comment (SuperSandro2000) [inline nixos/modules/services/networking/xray.nix:29]:
```suggestion
        type = types.nullOr types.path;
```

File: pkgs/tools/networking/xray/default.nix (hunk 2 above)
Comment (iopq) [reply pkgs/tools/networking/xray/default.nix:34]: I think it might be fine, but I haven't tested it yet

Comment (SuperSandro2000) [reply pkgs/tools/networking/xray/default.nix:34]: Please test it

File: pkgs/tools/networking/xray/default.nix [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    install -Dm755 xray -t $out/bin
+
+    for file in $out/bin/*; do
+      makeWrapper "$file" "$out/bin/$(basename "$file")" \
+        --set-default XRAY_LOCATION_ASSET ${assetsDrv}/share/xray
+    done
+  '';
+
+  meta = {
+    description = "Xray, Penetrates Everything. Also the best v2ray-core, with XTLS support and fully compatible configuration";
```
Comment (amjoseph-nixpkgs) [inline pkgs/tools/networking/xray/default.nix:46]: Could we provide a more useful description of what this software *does*?  I still don't know after reading this.  The current text sounds like some kind of tagline I'd find on an unhelpful blingy corporate website.

Comment (iopq) [reply pkgs/tools/networking/xray/default.nix]: The software description has been updated

File: pkgs/tools/networking/xray/default.nix (hunk 2 above)
Comment (iopq) [reply pkgs/tools/networking/xray/default.nix]: Just did the necessary changes for this

**Comment by SuperSandro2000** [comment]
Please follow the [contributing guide](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#submitting-changes) when naming your commits.

**Pull request closed without merging**